### DISCUSSION TITLE
Enhancing Route Generation with Query Parameters

### DISCUSSION THREAD

**Discussion by Jagadish056** [laravel/framework#52703] [Ideas]
While [editing a row](https://alpine-ajax.js.org/examples/edit-row/), generating URLs with query parameters is a common task, particularly when dealing with pagination, sorting, and filtering. However, manually appending query parameters to routes can make the code cumbersome. For example:
```
<a href="{{ route('contacts.edit', ['contact' => $contact] + request()->query()) }}" x-target="contact_table>Edit</a>
```

To improve code readability and streamline this process, I suggest implementing a feature that automatically appends the current query parameters to the generated URL.

**Proposed Solution:**

One possible approach is to extend the `route()` helper to optionally include existing query parameters. If modifying `route()` isn’t feasible, we could introduce a new helper function, `routeWithQuery()`, to handle this.

Here’s a potential implementation:
```
if (!function_exists('routeWithQuery')) {
    /**
     * Generate a URL for the given route and append the current query parameters.
     *
     * @return string
     */
    function routeWithQuery(...$args)
    {
        return url()->query(route(...$args), request()->query());
    }
}
```
With this helper, the code can be simplified to:
```
<a href="{{ routeWithQuery('contacts.edit', $contact) }}" x-target="contact_table">Edit</a>
```

**Comment by shaedrich** [comment]
I like the overall idea, but the naming doesn't quite cut it for me as it doesn't mirror your intent clearly and unambiguously. `routeWithAdditionalQueryParams()` would be more precise, but unarguably long.

Maybe, the problem is that the method always returns a string. If it instead returned a stringable class instance (which would still work the same in blade then), additional calls to methods of said class could be chained until the route is constructed:

```
class Route
{
    public function __construct(private string $uri) {}

    public function setQuery(array $queryParams)
    {
        $queryString = '?'.implode('&', array_map(function($key, $value) {
            return "$key=$value";
        }, array_keys($queryParams), array_values($queryParams)));
        if (str_contains($this->uri, '?')) {
            $this->uri = preg_replace('/\?(.*)$/', $queryString, $this->uri);
        } else {
            $this->uri .= $queryString;
        }
        
        return $this;
    }

    public function addQuery(array $queryParams)
    {
        $queryString = implode('&', array_map(function($key, $value) {
            return "$key=$value";
        }, array_keys($queryParams), array_values($queryParams)));
        if (str_contains($this->uri, '?')) {
            $this->uri .= '&'.$queryString;
        } else {
            $this->uri .= '?'.$queryString;
        }
        
        return $this;
    }

    public function __toString()
    {
        return $this->uri;
    }
}
```
<small>_(of course, this is just an example and not intended to be a real implementation)_</small>
